### PR TITLE
feat(ingest): Add config option to set Bigquery credential in source config

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist
- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: metadata-ingestion/source_docs/bigquery.md [hunk 1]
```diff
@@ -6,6 +6,63 @@ For context on getting started with ingestion, check out our [metadata ingestion
...
+}
+```
+3. Here you have two options:
+   Either point to it using an environment variable:
+       $ export GOOGLE_APPLICATION_CREDENTIALS="/path/to/keyfile.json"
+
+   *or*
+
+   Set credential config in your source based on the credential json file like this:
+```yml
```
Comment (shirshanka) [inline metadata-ingestion/source_docs/bigquery.md:57]: This is amazing! Thank you.

File: metadata-ingestion/source_docs/bigquery.md [hunk 2]
```diff
@@ -6,6 +6,63 @@ For context on getting started with ingestion, check out our [metadata ingestion
...
+and assign the previously created role to this service account
+2. Download a service account JSON keyfile:
+   The credential file looks like this:
+```json
+{
+  "type": "service_account",
+  "project_id": "project-id-1234567",
+  "private_key_id": "d0121d0000882411234e11166c6aaa23ed5d74e0",
+  "private_key": "-----BEGIN PRIVATE KEY-----\nMIIyourkey\n-----END PRIVATE KEY-----",
+  "client_email": "test@suppproject-id-1234567.iam.gserviceaccount.com",
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/bigquery.md:42]: Amazing!!!

File: metadata-ingestion/source_docs/bigquery.md [hunk 3]
```diff
@@ -6,6 +6,63 @@ For context on getting started with ingestion, check out our [metadata ingestion
...
+   bigquery.tables.getData # Needs for profiling
+   bigquery.tables.list
+   logging.logEntries.list # Needs for lineage generation
+   resourcemanager.projects.get
+```
+### Create a service account:
+
+1. Setup a ServiceAccount (https://cloud.google.com/iam/docs/creating-managing-service-accounts#iam-service-accounts-create-console)
+and assign the previously created role to this service account
+2. Download a service account JSON keyfile:
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/bigquery.md:34]: There are 2 ":" in this line :)

File: metadata-ingestion/source_docs/bigquery.md [hunk 4]
```diff
@@ -6,6 +6,63 @@ For context on getting started with ingestion, check out our [metadata ingestion
...
+  "private_key": "-----BEGIN PRIVATE KEY-----\nMIIyourkey\n-----END PRIVATE KEY-----",
+  "client_email": "test@suppproject-id-1234567.iam.gserviceaccount.com",
+  "client_id": "113545814931671546333",
+  "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+  "token_uri": "https://oauth2.googleapis.com/token",
+  "auth_provider_x509_cert_url": "https://www.googleapis.com/oauth2/v1/certs",
+  "client_x509_cert_url": "https://www.googleapis.com/robot/v1/metadata/x509/test%suppproject-id-1234567.iam.gserviceaccount.com"
+}
+```
+3. Here you have two options:
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/bigquery.md:50]: Let's rephrase as: 

To provide credentials to the source, you can either 

1. Set an environment variable
2. Provide credential fields in the source recipe

File: metadata-ingestion/source_docs/bigquery.md [hunk 5]
```diff
@@ -6,6 +6,63 @@ For context on getting started with ingestion, check out our [metadata ingestion
...
+   Either point to it using an environment variable:
+       $ export GOOGLE_APPLICATION_CREDENTIALS="/path/to/keyfile.json"
+
+   *or*
+
+   Set credential config in your source based on the credential json file like this:
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/bigquery.md:56]: Set credential config in your source based on the credential json file. For example:

File: metadata-ingestion/source_docs/bigquery.md [hunk 6]
```diff
@@ -6,6 +6,63 @@ For context on getting started with ingestion, check out our [metadata ingestion
...
+```yml
+     credential:
```
Comment (jjoyce0510) [inline metadata-ingestion/source_docs/bigquery.md:58]: Wondering if this should also be in the example recipe :)

File: metadata-ingestion/src/datahub/ingestion/source/sql/bigquery.py [hunk 7]
```diff
@@ -140,20 +144,59 @@ def get_view_definition(self, connection, view_name, schema=None, **kw):
...
 
+    def __init__(self, **data: Any):
+        super().__init__(**data)
+
+        if self.credential:
+            self.credentials_path = create_credential_temp_file(self.credential)
+            logger.debug(
+                f"Creating temporary credential file at {self.credentials_path}"
+            )
+            os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = self.credentials_path
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/sql/bigquery.py:198]: Is this strictly required? It is strange we need to set an env variable

File: metadata-ingestion/src/datahub/ingestion/source/sql/bigquery.py [hunk 8]
```diff
@@ -479,3 +522,11 @@ def standardize_schema_table_names(
         if segments[0] != schema:
             raise ValueError(f"schema {schema} does not match table {entity}")
         return segments[0], segments[1]
+
+    # We can't use close as it is not called if the ingestion is not successful
+    def __del__(self):
+        if self.config.credentials_path:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/sql/bigquery.py:528]: Nice!

File: metadata-ingestion/tests/unit/test_bigquery_source.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    )
+
+    try:
+
+        assert config.get_sql_alchemy_url() == "bigquery://test-project"
+        assert config.credentials_path
+
+        with open(config.credentials_path) as jsonFile:
+            json_credential = json.load(jsonFile)
+            jsonFile.close()
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/unit/test_bigquery_source.py:56]: Last q: Did you try setting the credentials using the config method manually?

Comment (treff7es) [reply metadata-ingestion/tests/unit/test_bigquery_source.py:56]: Yes, I tried it out a couple of times on my computer.

File: metadata-ingestion/src/datahub/ingestion/source/sql/bigquery.py (hunk 7 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/sql/bigquery.py:198]: We are using various APIs (sqlalchemy, BigQueryClient (https://cloud.google.com/bigquery/docs/reference/libraries#setting_up_authentication) and GCPLoggingClient (https://cloud.google.com/logging/docs/reference/libraries#setting_up_authentication), and with this way, we can authenticate with all of these.
I think it is simple and safe as this environment variable only lives during the life of the python app but works for all the APIs.

File: metadata-ingestion/source_docs/bigquery.md (hunk 6 above)
Comment (treff7es) [reply metadata-ingestion/source_docs/bigquery.md:58]: Great idea, I will add it.

File: metadata-ingestion/examples/recipes/azure_ad_to_datahub.yml [hunk 10]
```diff
@@ -19,4 +19,4 @@ source:
 sink:
   type: "datahub-rest"
   config:
-    server: "https://autotrader.acryl.io/gms"
\ No newline at end of file
+    server: "http://localhost:8080"
```
Comment (jjoyce0510) [inline metadata-ingestion/examples/recipes/azure_ad_to_datahub.yml:22]: Thank you!

**Review by jjoyce0510 — APPROVED** [review]
LGTM. Thank you so much @treff7es !

**Review by shirshanka — APPROVED** [review]
LGTM

**Pull request merged**